llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -610,13 +610,59 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
       // unknown users
     }
 
+    if (IsSelfRecursive && isa<CallBase>(V)) {
+      Value *Ptr = dyn_cast<Value>(U);
```

Comment:
unchecked dyn_cast, which is also redundant with the isa. Just dyn_cast to CallBase in the first place? You also need to check this is actually the callee argument

Reply:
U is the Use, and not the User, so it's not redundant with the isa in the if condition. Am I missing something ?

Reply:
> You also need to check this is actually the callee argument

That's being checked below, right ?

```c
      if (PtrArg != Arg)
        return false;
```